### PR TITLE
Issue/2630 display grouped products

### PR DESCRIPTION
**Pull request by anitaa1990**
Part 1/3 of #2630. This PR adds logic to display products associated with grouped products.

#### Changes
- Added logic to display the product count for a grouped product, if the count > 0.
- Created a new fragment, module and View Model to handle the changes needed to show only the products for a specific grouped product.
- Split the `ProductListViewHolder` and `ProductListDiffUtil` to separate classes so that they can be reused by `GroupedProductListAdapter` as well as `ProductListAdapter`

#### Screenshots
. . . 

#### Testing
- Click on a grouped product with atleast 2 associated products.
- Notice the grouped products section in the Product Detail screen.
- Click on the grouped products section and notice the list of products associated with this product, displayed correctly.
- Click on the `delete` icon for one of the items and verify:
  - that the "Done" menu button is displayed correctly.
  - that clicking on the back button displays the discard dialog.
- Click on the "Done" menu button and notice that 
  - the "Update" button is displayed correctly in the Product detail screen.
  - the grouped products section displays the correct count.
- Click on the "Update" menu button and verify that the product is updated correctly.

#### Notes
- ~This PR is in draft till this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1668) can be reviewed and merged.~
- The option to add a grouped product will take place in a subsequent PR.

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Review by AmandaRiu — CHANGES_REQUESTED** [review]
@anitaa1990 Tried to test, but am getting this crash:

```
2020-08-17 16:57:01.147 30517-30517/com.woocommerce.android.dev E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.woocommerce.android.dev, PID: 30517
    java.lang.IllegalStateException: Gson().fromJson(groupedP… JsonElement::class.java) must not be null
        at org.wordpress.android.fluxc.model.WCProductModel.getGroupedProductIds(WCProductModel.kt:245)
        at com.woocommerce.android.model.ProductKt.toAppModel(Product.kt:548)
        at com.woocommerce.android.ui.products.ProductListRepository.getProductList(ProductListRepository.kt:141)
        at com.woocommerce.android.ui.products.ProductListViewModel$loadProducts$2.invokeSuspend(ProductListViewModel.kt:195)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

**Comment by anitaa1990** [comment]
Hey @AmandaRiu!

> Tried to test, but am getting this crash:

I'm not actually able to reproduce the crash. I tested with a branch new grouped product as well as existing product. But looking at the error, it seems to be that `groupedProductIds` is null (which shouldn't happen) but I added logic in FluxC to verify if the `groupedProductIds` is empty before parsing it. I opened a [PR in FluxC](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1670) as well.

Ready for another round 👍 Let me know if you are able to reproduce the crash. I would also like to reproduce, if possible so any testing steps would be helpful as well.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListRepository.kt [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            emptyList()
+        }
+    }
+
+    @SuppressWarnings("unused")
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    fun onProductChanged(event: OnProductChanged) {
+        if (event.causeOfChange == FETCH_PRODUCTS) {
+            if (event.isError) {
+                // TODO: add tracking event
```
Comment (AmandaRiu) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListRepository.kt:94]: Just noting there is a TODO here for tracks events. Will that be handled in a separate PR? Same with a couple lines down.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt [hunk 2]
```diff
@@ -469,6 +471,26 @@ class ProductDetailCardBuilder(
...
+        val groupedProductsSize = this.groupedProductIds.size
+        return if (FeatureFlag.PRODUCT_RELEASE_M3.isEnabled() && groupedProductsSize > 0) {
+            val groupedProductResourceId = if (groupedProductsSize == 1) R.string.grouped_products_single
+            else R.string.grouped_products_count
+            ComplexProperty(
+                R.string.grouped_products,
+                resources.getString(groupedProductResourceId, groupedProductsSize),
+                R.drawable.ic_widgets
+            ) {
+                // TODO: add click event
```
Comment (AmandaRiu) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailCardBuilder.kt:484]: Note: tracks event TODO.

**Review by AmandaRiu — APPROVED** [review]
Tested and looks good! 👍 Nice job Anitaa!  :shipit:

**Pull request merged**